File: apps/server/src/healthee/read/vo2max.py
"""The VO2max payload — the estimate, the instrument behind it, and why there isn't one.

## The number is tiered, and this file does NOT decide the tier (#117)

``vo2max_estimate`` is one metric with three instruments in a fixed order — graded GPS
fit, then the %HRR reserve inversion, then the Jurca non-exercise model. That order is
applied once, when the day's row is written (``derive/vo2max_tier.py``), and this payload
reads the row. It never re-tiers: a read-time choice would be a second definition of the
metric, and the day's row would then mean different things depending on who asked.

What this file owes the owner is the OTHER half of [[hr_reserve_vo2max]] Directive 4 —
"state which one produced the value". ``method``, ``method_caveat``, ``see_source`` and
``research_notes`` all move with the instrument, so a number that came from a run last
week and a questionnaire this week cannot look like the same measurement twice.


Split out of ``read/fitness.py`` when the freshness gate below pushed that file past
the 400-line limit. It was the right home anyway: everything here answers "what do we
actually know about this owner's aerobic fitness TODAY", which is a different question
from load, MVPA and strength (standards §1: a file has one reason to change).

## Withheld must READ as withheld (2026-07-31)

``derive/vo2max.py`` withholds the Jurca estimate — writes NO row — when its inputs
cannot carry it: no profile or logged weight, fewer than 3 resting-HR days, an RHR
median outside Jurca's validated 40-100 bpm, or a 7-day RHR MAD above 8 bpm. The note
is explicit that this is a decision, not a gap: "Skip the derivation (show
'Insufficient data') ... never write a wrong value."

This payload used to read the last 95 days and take ``rows[-1]``. So on a withheld day
it rendered an OLDER day's estimate as the hero number. It carried ``as_of_date``, so
it was not fabrication — but a date in a field the UI may not render does not undo a
confident current-looking number, and the withhold gate exists precisely to say "we
don't have enough to tell you today". In user terms the gate was defeated.

The fix is structural, not a label: **an estimate is only reported when the newest row
is the owner's TODAY.** Otherwise ``estimate`` and ``as_of_date`` are both ``null``,
``data_confidence`` is ``insufficient_data`` (the outcome ledger's exact vocabulary —
``challenges/ledger.py``), and a ``withheld`` block names the reason and what it would
take to fix, in the shape ``data_health`` uses for a dead feed. A UI that renders the
hero number gets nothing to render; there is no "dated field" to miss.

The reason is RECOMPUTED, not persisted: ``derive.vo2max.withhold_reason_for_day``
re-runs the same gate over the same inputs, which are still in the database. No
migration, no backfill, and it works for rows written before this existed.

What is deliberately KEPT when the estimate is withheld: ``trend_90d`` (the note calls
the trend the trustworthy signal, and a trend that ends before today is honest as long
as nothing claims it ends now), ``median_for_age`` (a population fact about the owner's
age and sex, not a claim about them), and the last estimate itself — inside the
``withheld`` block, where it cannot be mistaken for today's.
"""

from __future__ import annotations

from datetime import date
from uuid import UUID

from healthee.analytics.reference_scales import vo2max_median_for
from healthee.core.tenancy import AS_OF_DAY_SQL, reference_day
from healthee.derive._common import Cur
from healthee.derive.freshness import withheld_block
from healthee.derive.vo2max import (
    METHOD_JURCA,
    WITHHOLD_MESSAGES,
    out_of_range_inputs,
)
from healthee.derive.vo2max_reserve import METHOD_RESERVE
from healthee.derive.vo2max_submax import METHOD_GRADED
from healthee.derive.vo2max_tier import estimate_unavailable_reason, method_of

_WINDOW_DAYS = 95  # the trend window; ~90 days of trend plus slack

# What each instrument's number is worth, in the second person. The sentences state a
# LIMIT rather than a confidence score, because the three methods do not differ in
# precision so much as in what they assume: the graded fit measures this owner's own
# VO2-HR line, the reserve inversion assumes a population equivalence that the largest
# study of it rejects ([[hr_reserve_vo2max]]), and the non-exercise model measures no
# exertion at all ([[non_exercise_vo2max]]).
#
# One of these ships with EVERY number this payload reports, because
# [[hr_reserve_vo2max]] Directive 4 is explicit that the instrument must be stated — an
# owner whose number comes from a run one week and a questionnaire the next has to be able
# to see that, or a change of instrument reads as a change in them.
_METHOD_CAVEATS = {
    METHOD_GRADED: (
        "Measured from a recorded session, by how your heart rate tracked your workload "
        "across it — the most direct of the three methods we have, because it measures "
        "the relationship on you rather than assuming it."
    ),
    METHOD_RESERVE: (
        "Measured from a recorded RUN, by how far into your heart-rate range you were, "
        "using a population relationship rather than one measured on you. It only runs on "
        "running, and it most likely reads a little LOW — the published bias in that "
        "relationship under-states fitness by roughly 2-3 mL/kg/min."
    ),
    METHOD_JURCA: (
        "Estimated without any exertion, from your age, sex, BMI, resting heart rate and "
        "your own answer about how much deliberate exercise you do — the fallback we use "
        "when no recent recorded session can measure it. It is the noisiest of the three."
    ),
}

# [[hr_reserve_vo2max]] D6 and [[submaximal_vo2max]] D5: "report the median across
# sessions … never one session as a fact". Inside the freshness horizon there is usually
# one session or none, so refusing at n = 1 would leave the measured tier permanently
# inert. The number ships and the fact-hood does not — this sentence is what makes the
# difference visible to the owner rather than only to the code.
_SINGLE_SESSION_CAVEAT = (
    " It comes from one session, not a run of them, so read it as that session rather "
    "than as a settled level."
)

# Which note licenses the number in front of the owner. The fitness↔mortality note is
# common to all three because it is what makes any VO₂max worth showing.
_METHOD_NOTES = {
    METHOD_GRADED: ["vo2max_fitness_mortality", "submaximal_vo2max"],
    METHOD_RESERVE: ["vo2max_fitness_mortality", "hr_reserve_vo2max"],
    METHOD_JURCA: ["vo2max_fitness_mortality", "non_exercise_vo2max"],
}


def method_caveat(method: str, n_sessions: int | None) -> str | None:
    """The instrument's sentence, plus the single-session qualifier when it applies."""
    base = _METHOD_CAVEATS.get(method)
    if base is None:
        return None
    return base + (_SINGLE_SESSION_CAVEAT if n_sessions == 1 else "")


def vo2max_payload(cur: Cur, user_id: UUID, tz: str, day: date | None = None) -> dict | None:
    """The tiered VO2max AS OF ``day`` + 90-day trend + the session record behind it.

    ``None`` only when the owner has no estimate at all in the window — genuinely
    nothing to say. When there IS history but the day has no estimate, the payload is
    returned with ``estimate: None`` and a ``withheld`` block: "we can't tell you
    for that day, and here is what we'd need" is a real answer, absence is not.
    [[vo2max]] (Mandsager 2018); derivation [[non_exercise_vo2max]].

    **This is the "latest" ``docs/AS_OF_DAY.md`` names first.** ``rows[-1]`` used to be
    the newest estimate the owner had; as of ``day`` it is the newest with ``day <=
    day``, because the window itself now closes there. A June answer that inherited an
    August measurement would be the future leak in its purest form — the same value
    dated wrongly, which is exactly what the stale-as-current gate below refuses in the
    other direction.
    """
    as_of = reference_day(day, tz)
    rows = _window(cur, user_id, as_of, "vo2max_estimate")
    if not rows:
        return None
    last_day, last_value, flags = rows[-1][0], float(rows[-1][1]), (rows[-1][2] or {})
    withheld = _withheld_block(cur, user_id, tz, as_of, last_day, last_value)
    age = int(flags.get("age_years") or 0)
    sex = str(flags.get("sex") or "male")
    median_ref = vo2max_median_for(age, sex) if age else None
    estimate = None if withheld else round(last_value, 1)
    # Rows written before #117 carry no ``method`` and are all Jurca. That default lives in
    # ``vo2max_tier.method_of`` because the coach's pivot resolves the same rows (#120) and
    # two answers to "what produced this row" is D4 half-kept.
    method = method_of(flags.get("method"))
    n_sessions = flags.get("n_sessions")
    return {
        "submax": _submax_block(cur, user_id, as_of),
        "estimate": estimate,
        "data_confidence": "insufficient_data" if withheld else "ok",
        "withheld": withheld,
        # WHICH INSTRUMENT produced the number above, on the wire, always. Paired with
        # `estimate`: null together, because naming the instrument behind a number we have
        # just declined to report would describe something the owner is not being shown.
        "method": None if withheld else method,
        "method_caveat": None if withheld else method_caveat(method, n_sessions),
        "measured_as_of": None if withheld else flags.get("measured_as_of"),
        "n_sessions": None if withheld else n_sessions,
        # The ± band, in the reporting instrument's own units of error — a MAPE for the
        # graded fit, a modelled SD for the reserve inversion, an SEE for Jurca.
        # ``see_source`` says WHICH, so a percentage error cannot be read as a standard
        # error of estimate. Older rows carry Jurca's SEE and nothing else.
        "see_source": flags.get("see_source"),
        "see_ml_kg_min": float(flags.get("see_ml_kg_min", 5.6)),
        # Paired with `estimate`: both describe today's number, so both are null when
        # there isn't one. The last day that DID have one lives in `withheld`.
        "as_of_date": None if withheld else last_day.isoformat(),
        "age_years": age,
        "sex": sex,
        "median_for_age": median_ref,
        "delta_from_median": _delta(estimate, median_ref),
        "trend_90d": [{"date": d.isoformat(), "value": round(float(v), 1)} for d, v, _ in rows],
        # v2 flag names: rhr_med_7d←rhr_med, pa_score←srpa; weekly_mvpa_min not
        # stored in v2 vo2max flags → null (documented WP7 note).
        "inputs": {
            "bmi": flags.get("bmi"),
            "rhr_med_7d": flags.get("rhr_med"),
            "weekly_mvpa_min": None,
            "pa_score": flags.get("srpa"),
        },
        # Directive 5 of [[non_exercise_vo2max]]: an estimate computed outside the
        # range the model was validated on says so. Recomputed from the inputs the
        # row already stores (no schema, no backfill), so it also covers rows
        # written before the flag existed. Empty list = every input in range.
        "out_of_range_inputs": out_of_range_inputs(age or None, flags.get("bmi")),
        # The notes that license THIS number, which depends on which instrument read it.
        "research_notes": _METHOD_NOTES[method],
    }


def _withheld_block(
    cur: Cur, user_id: UUID, tz: str, as_of: date, last_day: date, last_value: float
) -> dict | None:
    """Why there is no estimate for ``as_of``, or None when that day has one.

    A row for any day other than ``as_of`` does not make ``as_of``'s estimate exist, and
    ``derive.vo2max_tier.estimate_unavailable_reason`` is the check that makes that
    structural: a withheld day cannot "resurrect" on the next one just because some row
    survives inside the 95-day window. That rule lives beside the gate it belongs to
    because biological age needs the identical answer (standards §Duplication).

    The gate already took its reference day as an argument, so answering for a past day
    needed nothing added to it — only that this caller stop hardwiring the wall clock.
    Its own withhold checks (profile, SR-PA, the resting-HR window) are re-run AS OF that
    day, which is the point: a day whose inputs could not carry a number then must not be
    told they can now.

    Two queries, and only on the days that need them — a fresh estimate short-circuits
    before the gate is re-run, so the common path costs nothing extra.
    """
    reason = estimate_unavailable_reason(cur, user_id, tz, as_of, last_day)
    if reason is None:
        return None
    return withheld_block(
        reason, WITHHOLD_MESSAGES[reason], as_of, last_day, last_estimate=round(last_value, 1)
    )


def _delta(estimate: float | None, median_ref: float | None) -> float | None:
    """Estimate vs the age/sex population median — a claim about NOW, so it needs a
    current estimate. Null whenever the estimate is withheld."""
    if estimate is None or not median_ref:
        return None
    return round(estimate - median_ref, 1)


def _window(cur: Cur, user_id: UUID, as_of: date, metric: str) -> list[tuple]:
    """(day, value, flags) for one metric over the trend window ENDING at ``as_of``.

    Both bounds, and the closing one is the load-bearing half: ``rows[-1]`` is read as
    "the estimate that speaks for this day" by every caller, so a window open at the top
    would hand a past day a value measured after it.
    """
    cur.execute(
        "SELECT day, value, flags FROM derived_daily "
        "WHERE user_id = %s AND metric = %s "
        f"AND day >= ({AS_OF_DAY_SQL} - %s::int) AND day <= {AS_OF_DAY_SQL} ORDER BY day",
        (user_id, metric, as_of, _WINDOW_DAYS, as_of),
    )
    return cur.fetchall()


def _submax_block(cur: Cur, user_id: UUID, as_of: date) -> dict | None:
    """The SESSION RECORD beneath the estimate: what each recorded effort measured.

    ``vo2max_submax`` is one row per day that carried a scoreable session — an
    observation, not a second answer to "what is this person's VO₂max". Since #117 that
    question has exactly one answer, ``estimate`` above, and on any day a session inside
    the freshness horizon exists these numbers ARE that answer rather than a rival to it
    (``derive/vo2max_tier.py``). Two metrics both meaning the owner's VO₂max is what
    CLAUDE.md's canonical-definition rule forbids; a metric and its observations is not
    that, in the same way ``rhr_daily`` and the heart-rate samples under it are not.

    ``vs_jurca`` lived here until #117 and is GONE with the tiering: it was a difference
    between two definitions of one quantity, which is the object that should not have
    existed. The comparison an owner can still make honestly is the one against the
    population reference (``delta_from_median``), and the instrument behind today's number
    is now stated at the top level instead of implied by a gap.

    ``last_method`` records which instrument read the newest session, so a session-to-
    session move that is really an instrument change does not read as a fitness change
    (#114). Rows written before #114 carry no ``method``; they are all graded fits, which
    is what ``METHOD_GRADED`` defaults them to.
    """
    rows = _window(cur, user_id, as_of, "vo2max_submax")
    if not rows:
        return None
    s_day, s_val, s_flags = rows[-1][0], float(rows[-1][1]), (rows[-1][2] or {})
    method = str(s_flags.get("method") or METHOD_GRADED)
    return {
        "latest": round(s_val, 1),
        # There was a ``median`` here, over every session in the window regardless of
        # which instrument read it. It went with ``vs_jurca`` in #117: a median over mixed
        # instruments IS the average [[hr_reserve_vo2max]] D4 forbids whenever the count
        # is even, and the only thing it fed was the comparison that no longer exists. The
        # median that IS reported is the tier's, over one instrument, and it is the
        # ``estimate`` at the top of this payload.
        "n_sessions": len(rows),
        "as_of_date": s_day.isoformat(),
        "last_r2": s_flags.get("r2"),
        "last_speed_kmh": s_flags.get("speed_kmh"),
        "last_method": method,
        "method_caveat": _METHOD_CAVEATS.get(method),
        "trend": [{"date": d.isoformat(), "value": round(float(v), 1)} for d, v, _ in rows],
    }

File: apps/mobile/lib/data/models/vo2max.dart
/// The VO₂max payload — the estimate, and which instrument produced it.
///
/// The mobile half of `read/vo2max.py`. Two of its rules are structural here
/// rather than advisory:
///
///   * **The instrument is not optional.** [[hr_reserve_vo2max]] Directive 4
///     requires that the method be stated wherever the number is, because "an
///     owner whose number comes from a run one week and a questionnaire the next
///     has to be able to see that, or a change of instrument reads as a change in
///     them". [method] and [methodCaveat] are therefore non-null on [Vo2max] —
///     you cannot construct this object without them, so no screen can render the
///     number and omit its instrument.
///   * **Withheld is not a zero.** The server nulls `estimate` and attaches a
///     `withheld` block; [Vo2max.maybe] returns null in exactly that case, and the
///     envelope turns it into a `Withheld<Vo2max>`. There is no `Vo2max` object
///     with a null estimate to accidentally render.
library;

import 'package:healthee/data/honesty/disclosure.dart';
import 'package:healthee/data/models/trend_point.dart';
import 'package:meta/meta.dart';

/// The inputs behind a non-exercise estimate, as the row stored them.
@immutable
class Vo2maxInputs {
  /// Builds an input set. Every field is nullable — see [Vo2max.inputs].
  const Vo2maxInputs({
    required this.bmi,
    required this.restingHrMedian7d,
    required this.weeklyMvpaMin,
    required this.physicalActivityScore,
  });

  /// Parses `vo2max.inputs`.
  factory Vo2maxInputs.fromJson(Map<String, Object?> json) {
    double? number(String key) => (json[key] as num?)?.toDouble();
    return Vo2maxInputs(
      bmi: number('bmi'),
      restingHrMedian7d: number('rhr_med_7d'),
      weeklyMvpaMin: number('weekly_mvpa_min'),
      physicalActivityScore: number('pa_score'),
    );
  }

  /// Body mass index at the time the row was written.
  final double? bmi;

  /// Seven-day median resting heart rate.
  final double? restingHrMedian7d;

  /// Moderate-to-vigorous minutes that week.
  final double? weeklyMvpaMin;

  /// Jurca's self-reported physical-activity score, 0–7.
  final double? physicalActivityScore;
}

/// `vo2max.submax` — the fit behind a session-measured estimate.
///
/// **Not a second estimate.** `derive/vo2max_tier.py` writes ONE VO₂max, and
/// this block describes the session the graded tier read it off: how well the
/// heart-rate/workload line fitted, and the speed the fit was taken at. It is
/// parsed so the fitness screen can answer *"which instrument produced it?"*
/// with the fit's own numbers instead of restating the estimate.
@immutable
class Vo2maxSubmax {
  /// Builds the block. Prefer [Vo2maxSubmax.maybe].
  const Vo2maxSubmax({
    required this.lastMethod,
    required this.lastR2,
    required this.lastSpeedKmh,
    required this.asOfDate,
  });

  /// Parses `submax`, or null when it holds nothing worth drawing.
  static Vo2maxSubmax? maybe(Map<String, Object?> json) {
    final block = Vo2maxSubmax(
      lastMethod: json['last_method'] as String?,
      lastR2: (json['last_r2'] as num?)?.toDouble(),
      lastSpeedKmh: (json['last_speed_kmh'] as num?)?.toDouble(),
      asOfDate: json['as_of_date'] as String?,
    );
    return block.isEmpty ? null : block;
  }

  /// The instrument the last scoreable session was read with.
  final String? lastMethod;

  /// How well that session's heart-rate/workload line fitted, 0–1.
  final double? lastR2;

  /// The speed the fit was taken at.
  final double? lastSpeedKmh;

  /// The day that session was recorded.
  final String? asOfDate;

  /// Whether every field is absent — a block with nothing to say.
  bool get isEmpty =>
      lastMethod == null && lastR2 == null && lastSpeedKmh == null;
}

/// A reported VO₂max, with the instrument that read it and that instrument's error.
@immutable
class Vo2max {
  /// Builds an estimate. Prefer [Vo2max.maybe] via the honesty envelope.
  const Vo2max({
    required this.estimate,
    required this.method,
    required this.methodCaveat,
    required this.standardErrorMlKgMin,
    required this.standardErrorSource,
    required this.asOfDate,
    required this.measuredAsOf,
    required this.medianForAge,
    required this.deltaFromMedian,
    required this.sessionCount,
    required this.trend90d,
    required this.researchNotes,
    required this.ageYears,
    required this.sex,
    required this.inputs,
    this.submax,
  });

  /// Parses the payload when it carries a current estimate; null when it does not.
  ///
  /// The null return is what the honesty envelope turns into a `Withheld`. Note
  /// that `method` is read only on this path: the server sets it to null in
  /// lockstep with `estimate`, "because naming the instrument behind a number we
  /// have just declined to report would describe something the owner is not being
  /// shown".
  static Vo2max? maybe(Map<String, Object?> json) {
    final estimate = (json['estimate'] as num?)?.toDouble();
    if (estimate == null) {
      return null;
    }
    return Vo2max(
      estimate: estimate,
      method: json['method']! as String,
      methodCaveat: json['method_caveat']! as String,
      standardErrorMlKgMin: (json['see_ml_kg_min'] as num?)?.toDouble(),
      standardErrorSource: json['see_source'] as String?,
      asOfDate: json['as_of_date'] as String?,
      measuredAsOf: json['measured_as_of'] as String?,
      medianForAge: (json['median_for_age'] as num?)?.toDouble(),
      deltaFromMedian: (json['delta_from_median'] as num?)?.toDouble(),
      sessionCount: (json['n_sessions'] as num?)?.toInt(),
      trend90d: TrendPoint.listFrom(json['trend_90d']),
      researchNotes: _strings(json['research_notes']),
      ageYears: (json['age_years'] as num?)?.toInt(),
      sex: json['sex'] as String?,
      inputs: Vo2maxInputs.fromJson(
        json['inputs'] is Map<String, Object?>
            ? json['inputs']! as Map<String, Object?>
            : const <String, Object?>{},
      ),
      submax: json['submax'] is Map<String, Object?>
          ? Vo2maxSubmax.maybe(json['submax']! as Map<String, Object?>)
          : null,
    );
  }

  /// mL/kg/min.
  final double estimate;

  /// Which instrument produced it: `gps_graded`, `hr_reserve`, or
  /// `jurca_non_exercise`. Non-null by construction — see the library docstring.
  final String method;

  /// That instrument's limit, in the second person. Always rendered with the
  /// number; it is the honest half of the estimate.
  final String methodCaveat;

  /// [methodCaveat] as the disclosure it is, so the cards carry it the same way
  /// every other caveat on the app is carried.
  ///
  /// It arrives outside the `caveats` array — it is a property of the instrument
  /// rather than of this reading — but it is the same kind of claim and the same
  /// length: the live one is 568 characters, and printing it as body prose is
  /// what made Today's Fitness section an essay. Both VO₂max cards render it
  /// through `CaveatNote` now, which puts it one tap away instead of on the card.
  ///
  /// **The method itself stays ON the card.** [[hr_reserve_vo2max]] Directive 4
  /// requires the instrument to be named wherever the number is — that is the
  /// short "Read by …" line, not this paragraph, and moving the paragraph does
  /// not touch it.
  Disclosure get methodDisclosure =>
      Disclosure(reason: 'vo2max_method_caveat', message: methodCaveat);

  /// The ± band, in the reporting instrument's own units of error.
  ///
  /// `docs/APP_DESIGN.md` §3.1 and §3.3 both say "always show the ± band", which
  /// is why this is on the model rather than looked up per screen.
  final double? standardErrorMlKgMin;

  /// WHICH kind of error the band is — a MAPE, a modelled SD, or an SEE. Without
  /// it a percentage error can be misread as a standard error of estimate.
  final String? standardErrorSource;

  /// The day this estimate is a claim about, `YYYY-MM-DD`.
  final String? asOfDate;

  /// The day the MEASUREMENT behind it was recorded, `YYYY-MM-DD`.
  ///
  /// A different fact from [asOfDate] and that is the whole reason it is here.
  /// `derive/vo2max_tier.py` lets a graded session or a reserve inversion speak
  /// for up to `MEASURED_VO2MAX_MAX_AGE_DAYS` = 14 days, and the freshness
  /// argument that permits it (`derive/freshness.py`) rests on the owner being
  /// told which day it was measured. This was on the wire and unparsed, so a card
  /// read "as of today" over a run recorded a fortnight ago — the stale-as-current
  /// lie reached through the client rather than the server, which is why the
  /// server's own guard could not catch it.
  ///
  /// Rendered only when it DIFFERS from [asOfDate]; see [measuredEarlier].
  final String? measuredAsOf;

  /// The population median for the owner's age and sex — a fact about the
  /// reference group, not a claim about them.
  final double? medianForAge;

  /// Estimate minus [medianForAge]. Null when either is missing.
  final double? deltaFromMedian;

  /// How many recorded sessions are behind a measured estimate. 1 means the
  /// number is one session, not a settled level, and the caveat says so.
  final int? sessionCount;

  /// Up to 90 days of history. Kept even when the estimate is withheld — a trend
  /// that ends before today is honest as long as nothing claims it ends now.
  final List<TrendPoint> trend90d;

  /// The measurement day when it is NOT the day the estimate is offered for.
  ///
  /// Null when the two agree, so a screen that prints this prints nothing on the
  /// ordinary day and names the gap on the day there is one.
  String? get measuredEarlier =>
      measuredAsOf != null && measuredAsOf != asOfDate ? measuredAsOf : null;

  /// The notes that license this number in front of the owner.
  final List<String> researchNotes;

  /// The age the population median was looked up for. A fact about the reference
  /// group; legacy prints it beside [sex] in the module's header.
  final int? ageYears;

  /// `male` or `female`, as the profile holds it — again, the reference group.
  final String? sex;

  /// The four numbers Jurca's non-exercise model takes.
  ///
  /// Every one of them can be null, and `bmi` / `rhr_med_7d` / `pa_score` exist
  /// only on a row the Jurca tier wrote.
  ///
  /// `weekly_mvpa_min` used to be null on EVERY row — `read/vo2max.py` mapped it
  /// to `None` outright, beside a note saying v2 does not store it in the VO₂max
  /// flags. That was true and it was not a reason: the number is computed twenty
  /// lines away for `/api/activity.mvpa.week_min`, and both surfaces read it from
  /// one place now (`docs/BACKEND_GAPS_FROM_UI.md` B3). It is null only when
  /// there is genuinely no MVPA row to sum — never `0`, which would read as a
  /// measured week of stillness.
  final Vo2maxInputs inputs;

  /// The fit behind a session-measured estimate, when there was a session.
  final Vo2maxSubmax? submax;

  static List<String> _strings(Object? raw) {
    if (raw is! List) {
      return const [];
    }
    return [
      for (final entry in raw)
        if (entry is String) entry,
    ];
  }
}
